monai/apps/deepgrow/transforms.py

```diff
@@ -0,0 +1,623 @@
+        self.guidance = guidance
+        self.connected_regions = connected_regions
+
+    def randomize(self, data=None):
+        pass
+
+    def _apply(self, label, sid):
+        dimensions = 3 if len(label.shape) > 3 else 2
+        default_guidance = [-1] * (dimensions + 1)
+
+        dims = dimensions
+        if sid is not None and dimensions == 3:
+            dims = 2
+            label = label[0][sid][np.newaxis]  # Assume channel is first
+
+        label = (label > 0.5).astype(np.float32)
+        blobs_labels = measure.label(label.astype(int), background=0) if dims == 2 else label
+        assert np.max(blobs_labels) > 0, "Not a valid Label"
+
+        pos_guidance = []
+        for ridx in range(1, 2 if dims == 3 else self.connected_regions):
+            if dims == 2:
+                label = (blobs_labels == ridx).astype(np.float32)
+                if np.sum(label) == 0:
+                    pos_guidance.append(default_guidance)
+                    continue
+
+            distance = distance_transform_cdt(label).flatten()
+            probability = np.exp(distance) - 1.0
+
+            idx = np.where(label.flatten() > 0)[0]
+            seed = self.R.choice(idx, size=1, p=probability[idx] / np.sum(probability[idx]))
+            dst = distance[seed]
+
+            g = np.asarray(np.unravel_index(seed, label.shape)).transpose().tolist()[0]
+            g[0] = dst[0]
+            if dimensions == 2 or dims == 3:
+                pos_guidance.append(g)
+            else:
+                pos_guidance.append([g[0], sid, g[-2], g[-1]])
+
+        return np.asarray([pos_guidance, [default_guidance] * len(pos_guidance)])
+
+    def __call__(self, data):
+        sid = data.get(self.sid)
+        sids = data.get(self.sids)
+        if sids is not None:
+            if sid is None or sid not in sids:
+                sid = self.R.choice(sids, replace=False)
+        else:
+            sid = None
+        data[self.guidance] = self._apply(data[self.label], sid)
+        return data
+
+
+class AddGuidanceSignald(Transform):
+    def __init__(self, image="image", guidance="guidance", sigma=2, number_intensity_ch=1, batched=False):
+        self.image = image
+        self.guidance = guidance
+        self.sigma = sigma
+        self.number_intensity_ch = number_intensity_ch
+        self.batched = batched
+
+    def _get_signal(self, image, guidance):
+        dimensions = 3 if len(image.shape) > 3 else 2
+        guidance = guidance.tolist() if isinstance(guidance, np.ndarray) else guidance
+        if dimensions == 3:
+            signal = np.zeros((len(guidance), image.shape[-3], image.shape[-2], image.shape[-1]), dtype=np.float32)
+        else:
+            signal = np.zeros((len(guidance), image.shape[-2], image.shape[-1]), dtype=np.float32)
+
+        sshape = signal.shape
+        for i in range(len(guidance)):
+            for point in guidance[i]:
+                if np.any(np.asarray(point) < 0):
+                    continue
```

Comment:
Do we allow some point to has coordinates < 0?
Should we raise Error here?

Reply:
yes.. -1.. it's not an error.. it's to make +ve and -ve points array length in sync to batch them during training